File: python/data_src_dependent/graph_bacon_number/graph_bacon_number.py
from bridges.bridges import *
from bridges.graph_adj_list import *
from bridges.data_src_dependent.data_source import *

class Link():

    def __init__(self, nextVal, it = None):
        self.element = it
        self.next = nextVal

    def __call__(self):
        return self.next, self.element

    def next(self):
        return self.next

    def setNext(self, nextVal):
        self.next = nextVal
        return self.next

    def element(self):
        return self.element

    def setElement(self, it):
        self.element = it
        return self.element

class LQueue:

    def __init__(self, size = None):
        self.front = self.rear = Link(nextVal=None)
        self.size = 0


    def clear(self):
        self.__init__()

    def enqueue(self, it):
        self.rear.setNext(Link(nextVal=None, it=it))
        self.rear = self.rear.next
        self.size += 1

    def dequeue(self):
        assert self.size != 0, "queue is empty"
        it = self.front.next.element
        self.front.setNext(self.front.next.next)
        if self.front.next is None:
            self.rear = self.front
        self.size -= 1
        return it

    def frontValue(self):
        assert self.size != 0, "queue is empty"
        return self.front.next().element()

    def length(self):
        return self.size


def get_bacon_number(gr, src_actor, dest_actor, mark, dist, parent):
    #need to use queue, as we are soing a BFS search
    lq = LQueue()

    #add soruce actor to the queue
    lq.enqueue(src_actor)

    #inits
    mark[src_actor] = "visited"
    dist[src_actor] = 0
    parent[src_actor] = "none"

    #BFS Traversal
    while lq.length() > 0: #non empty queue

        vertex = lq.dequeue()

        #get adjacencey list of vertex
        sl_list = gr.get_adjacency_list(vertex)
        sle = sl_list
        while(sle is not None):
            #get destination vertex
            w = sle.value.tov

            # if unvisited, mark it as visited and add to the queue,
            # increment distance and point parent back to vertex
            if(mark[w] == ("unvisited")):
                mark[w] = "visited"
                lq.enqueue(w)

                #update the distance and parent values
                dist[w] = dist.get(vertex) + 1
                parent[w] = vertex

                # if we found the destination actor, then we are done
                #we want to highlight the actor and the path to the source
                # actor; we use the parent links to retrace (not source actor's
                # parent is "none" for us to stop
                if(w == dest_actor):#found it
                    #highlight the actor and path using
                    #the parent values and the hashmap
                    #to trace back the sequence of nodes
                    #to the source

                    #highlight the nodes and the links in the path from the source
                    #to the destination actor
                    p = dest_actor
                    while(p is not "none"):

                        #color the nodes in the path
                        #example gr.getVisualizer(key-val).setColor("red")

                        gr.get_visualizer(p).color = "red"
                        gr.get_visualizer(p).size = 50
                        gr.get_visualizer(p).opacity = 1.0

                        # next, color the link but check the parent is not "none", else you will
                        # get an exception
                        # example: gr.getLinkVisualizer(src-key, dest-key).setColor("red")
                        par = parent[p]
                        if(par is not "none"):
                            gr.get_link_visualizer(p, par).color = "red"
                            gr.get_link_visualizer(p, par).thickness = 10

                        p = par

                    return dist[dest_actor]
            sle = sle.next


def main():
    args = sys.argv[1:]

    # create the Bridges object, set credentials

    bridges = Bridges(int(args[0]), args[1], args[2])

    if len(args) > 3:
        bridges.connector.set_server(args[3])

    # set title for visualization
    bridges.set_title("Bacon Number: IMDB Actor-Movie Data")

    # use an adjacency list based graph
    gr = GraphAdjList()

    actor_list = get_actor_movie_imdb_data(1814)

    for k in range(len(actor_list)):

        # get the actor movie names
        actor = actor_list[k].actor
        movie = actor_list[k].movie

        # our graph needs to have a unique set of actors and movies
        # so create the actor and movie vertices only if they dont already
        # exist use an STL map to check for that

        vertices = gr.vertices

        # add actor if does not exist
        if (actor not in vertices):
            gr.add_vertex(actor, actor)

        # add movie if does not exist
        if (movie not in vertices):
            gr.add_vertex(movie, movie)

        # create the edge -- assumes no duplicate edges
        # undirected graph, edges go both ways
        gr.add_edge(actor, movie, 1)
        gr.add_edge(movie, actor, 1)

        # TO DO : Highlight "Cate_Blanchett" node and the movie nodes she is
        # connected to in "red" and do the same for "Kevin_Bacon_(I)" in "green"
        # specify colors by their names, "red", for example
        if (actor == ("Cate_Blanchett")):
            gr.get_link_visualizer(actor, movie).color = "orange"
            gr.get_visualizer(actor).color = "orange"
            gr.get_visualizer(movie).color = "orange"
        if (actor == ("Kevin_Bacon_(I)")):
            gr.get_link_visualizer(actor, movie).color = "green"
            gr.get_visualizer(actor).color = "green"
            gr.get_visualizer(movie).color = "green"

    # set the data structure handle and visualize the input graph
    bridges.set_data_structure(gr)
    bridges.visualize()

    mark = dict()
    parent = dict()
    dist = dict()

    # init maps for bacon number algorithm
    items = list(gr.vertices.keys())
    for i in range(len(items)):
        mark[items[i]] = "unvisited"
        parent[items[i]] = "none"
        dist[items[i]] = 0

    d = get_bacon_number(gr, "Kevin_Bacon_(I)", "Cate_Blanchett", mark, dist, parent)

    print("Bacon Number[Kevin Bacon to Cate Blanchett]: " + str(d))

    bridges.set_data_structure(gr)
    bridges.visualize()


if __name__ == "__main__":
    main()

File: java/data_src_dependent/graph_bacon_number/graph_bacon_number.java

import java.util.HashMap;
import java.util.Map.Entry;
import java.lang.String;
import java.util.List;
import java.util.LinkedList;

import bridges.connect.Bridges;
import bridges.connect.DataSource;
import bridges.base.SLelement;
import bridges.base.Element;
import bridges.base.Edge;
import bridges.base.GraphAdjList;
import bridges.base.GraphAdjListSimple;
import bridges.data_src_dependent.ActorMovieIMDB;



public class graph_bacon_number {
	public static void main(String[] args) throws Exception {
		
		// Initialize BRIDGES with your credentials

        Bridges bridges = new Bridges (Integer.parseInt(args[0]), args[1], args[2]);
        bridges.setServer (args[3]);
		// set title for visualization
		bridges.setTitle("Bacon Number: IMDB Actor-Movie Data");
		bridges.setDescription("Illustrates the shortest path between two actors: Kevin Bacon to Cary Grant");

		// use an adjacency list based graph
		GraphAdjListSimple<String> gr = new GraphAdjListSimple<>();

		// get the actor movie IMDB data through the BRIDGES API
		DataSource ds = bridges.getDataSource();
		List<ActorMovieIMDB> actor_list = ds.getActorMovieIMDBData (1814);

		String actor, movie;
		for (int k = 0; k < actor_list.size(); k++) {

			// get the actor and movie names
			actor = actor_list.get(k).getActor();
			movie = actor_list.get(k).getMovie();

			// our graph needs to have a unique set of actors and movies;
			// so create the actor and movie vertices only if they dont already
			// exist; use an STL map to check for that

			// first get the graph's vertex list
			HashMap<String, Element<String>> vertices = gr.getVertices();

			// add actor if it does not exist
			if (!vertices.containsKey(actor))
				gr.addVertex(actor, actor);

			// add movie if it does not exist
			if (!vertices.containsKey(movie))
				gr.addVertex(movie, movie);

			// create the edge -- assumes no duplicate edges 
			// undirected graph, edges go both ways
			gr.addEdge(actor, movie);
			gr.addEdge(movie, actor);

			// TO DO : Highlight "Cate_Blanchett" node and the movie nodes she is 
			// connected to in "red" and do the same for "Kevin_Bacon_(I)" in "green"
			// specify colors by their names, "red", for example
			String col = "";
			if (actor.equals("Charles_Grodin_(I)")) col = "orange"; 
			else if (actor.equals("Kevin_Bacon_(I)")) col = "green";
			else if (actor.equals( "Christopher_Lee_(I)")) col = "red";
			else if (actor.equals( "Morgan_Freeman_(I)")) col = "purple";
			else if (actor.equals( "James_Earl_Jones")) col = "yellow";
			else if (actor.equals( "Cary_Grant")) col = "cyan";
			else if (actor.equals( "Billy_Crystal")) col = "magenta";
			else if (actor.equals( "Cate_Blanchett")) col = "maroon";
			else col = "blue";

            gr.getLinkVisualizer (actor, movie).setColor (col);
//          gr.getLinkVisualizer (movie, actor).setColor (col);
            gr.getVisualizer (actor).setColor (col);
            gr.getVisualizer (movie).setColor (col);
		}

		//set the data structure handle, and visualize the input graph

		HashMap<String, String> mark = new HashMap<>(); 		//keeps track of visited nodes
		HashMap<String, String> parent = new HashMap<>(); 	//keeps track of parent nodes
		HashMap<String, Integer> dist = new HashMap<>(); 		// keeps track of distances

		// initialize maps for Bacon number algorithm
		for (Entry<String, Element<String> > v : gr.getVertices().entrySet() ) {
			mark.put(v.getKey(), "unvisited"); 
			parent.put(v.getKey(), "none");
			dist.put(v.getKey(), 0);
		}

		int d = getBaconNumber(gr, "Kevin_Bacon_(I)", "Cary_Grant", 
										mark, dist, parent);

		bridges.setDataStructure(gr);
		bridges.visualize();
	}

//
// Computes the Bacon Number of a an actor (#links that takes you from the 
// source actor to the destination actor. 
//
	public static int getBaconNumber (GraphAdjListSimple<String> 
					gr,   		// input graph 
					String src_actor, 						// source actor 
					String dest_actor,						// destination actor
					HashMap<String, String> mark, 	// mark array -- keep track of visited nodes
					HashMap<String, Integer> dist, 		// distances to src node 
					HashMap<String, String> parent) { 	// parent of node - for finding path 
															// back to source actor
	try {
		// Need to use a queue, as we are doing a BFS search
		LinkedList<String> lq = new LinkedList<String>();

		// add the source actor to the queue
		lq.addLast(src_actor);
		// initializations
		mark.put(src_actor, "visited");
		dist.put(src_actor,  0);
		parent.put(src_actor, "none");

		// BFS traversal
		while (lq.size() > 0) {  // non empty queue
			String vertex = (String) lq.removeFirst();

			// get adjacency list of vertex
			SLelement<Edge<String, String>> sl_list = gr.getAdjacencyList(vertex);
			for (SLelement<Edge<String, String>> sle = sl_list; sle != null; sle = 
												sle.getNext()){
				// get destination vertex
				String w = ((Edge<String, String>)sle.getValue()).getFrom();

				// if unvisited, mark it as visited and add to queue,
				// increment distance and point point parent back to vertex name
				if (mark.get(w).equals("unvisited")){
					mark.put(w, "visited");
					lq.addLast(w);
									// update the distance and parent values
					dist.put(w, dist.get(vertex)+1);
					parent.put(w, vertex);

					// if we found the destination actor, then we are done
					// we want to highlight the actor and the path to the source
					// actor; we use the parent links to retrace (note source actor's
					// parent is "none" for us to stop)
					if (w.equals(dest_actor)) {  // found it
											// highlight actor and path using
											// the parent values and the hashmap
											// to trace back the sequence of nodes
											// to the source node
							
						// optional, we are deemphasizing the graph nodes by adjusting opacity	
	/*
						Element<String> *el, *el2;
						for (auto& v : *gr.getVertices()) {
							el = (Element<String>*) v.second;
							el.setOpacity(0.5f);
						}
	*/
							
						// highlight the nodes and the links in the path from the source
						// to the destination actor
						String p = dest_actor;
						while (!p.equals("none")) {

							//  color the nodes in the path
							//  example: gr.getVisualizer(key-val).setColor("red")
							gr.getVisualizer(p).setColor("red");
							gr.getVisualizer(p).setSize(50);
							gr.getVisualizer(p).setOpacity(1.0f);

							// next, color the link, but check the parent is not "none", else you will
							// get an exception
							//  example: gr.getLinkVisualizer(src-key, dest-key).setColor("red")
							String par = parent.get(p);
							if (!par.equals("none")) {
								gr.getLinkVisualizer(p, par).setColor("red");
								gr.getLinkVisualizer(p, par).setThickness(10.);
							}	
							p = par;
						}
						// return the Bacon number of this pair of actors
						return dist.get(dest_actor);
					}
				}
			}
		}
	}
	catch (Exception exc) {
		// TODO
	}
		return -1;
	} // getBaconNumber
} // class BaconNumber
